### PR TITLE
Add truffle SVE implementation

### PR DESCRIPTION
**Pull request by ypicchi-arm**
This PR adds an implementation of truffle using SVE as per planned in https://github.com/VectorCamp/vectorscan/issues/190. It  speeds truffle's benchmark up by about 15% and pass the unit tests.

### CODE REVIEW

**Review by paul-szczepanek-arm — APPROVED** [review]

File: src/nfa/truffle_simd.hpp [hunk 1]
```diff
@@ -53,6 +59,166 @@ const SuperVector<S> blockSingleMask(SuperVector<S> shuf_mask_lo_highclear, Supe
...
+    svuint8_t wide_shuf_mask_lo_highclear = svld1(lane_pred_16, (uint8_t*) &shuf_mask_lo_highclear);
+    svuint8_t wide_shuf_mask_lo_highset = svld1(lane_pred_16, (uint8_t*) &shuf_mask_lo_highset);
+
+    const u8 *work_buffer = buf;
+    const u8 *ret_val;
+
+    DEBUG_PRINTF("start %p end %p \n", work_buffer, buf_end);
+    assert(work_buffer < buf_end);
+
+    __builtin_prefetch(work_buffer +   64);
```
Comment (markos) [inline src/nfa/truffle_simd.hpp]: I have started an prefetch optimization and I've found that in some cases I got up to 50% faster performance in the microbenchmarks: https://github.com/VectorCamp/vectorscan/commit/ab36731b9dd5c3ea26f4870654b258f159a85a66
Could you test these changes for SVE as well and confirm they are also valid in this PR?

Comment (markos) [reply src/nfa/truffle_simd.hpp]: Or I could just do those changes after this PR gets merged...

**Comment by markos** [comment]
@ypicchi-arm @paul-szczepanek-arm 

Unfortunately compilation fails for fat runtimes on Arm (all compilers tested), this needs to be fixed in order for the PR to be merged.

https://buildbot-ci.vectorcamp.gr/#/builders/7/builds/68

File: src/nfa/truffle_simd.hpp (hunk 1 above)
Comment (ypicchi-arm) [reply src/nfa/truffle_simd.hpp]: I did not observe any difference on my machine with the legacy and the new prefetch, so I set it the same way you did.
baseline (neon): Truffle: 5 matches, 16000 * 62500 iterations, total elapsed time = 0.308 s, average bandwidth = 4935.293 MB/s
patched with default prefetch: Truffle: 5 matches, 16000 * 62500 iterations, total elapsed time = 0.259 s, average bandwidth = 5860.817 MB/s
patched with new prefetch: Truffle: 5 matches, 16000 * 62500 iterations, total elapsed time = 0.259 s, average bandwidth = 5858.518 MB/s

**Review by paul-szczepanek-arm — APPROVED** [review]

**Comment by markos** [comment]
same failure, fat runtimes on arm build failure: https://buildbot-ci.vectorcamp.gr/#/builders/7/builds/69

I can look at it and fix it for this PR if you want, so we can merge it soon.

**Comment by paul-szczepanek-arm** [comment]
I can see the problem, sorry. Wait till tomorrow please, Yoan is away today - will fix tomorrow ASAP.

**Comment by markos** [comment]
Apologies, I just noticed, could you please resubmit this PR against develop branch? I really need to add some rules about PRs against master directly.

**Comment by paul-szczepanek-arm** [comment]
Sorry, I missed that, maybe you should put the information about the branches in the Contributing heading rather than Versioning in the readme.

**Comment by markos** [comment]
> Sorry, I missed that, maybe you should put the information about the branches in the Contributing heading rather than Versioning in the readme.

Indeed, I will fix that. Again apologies, I'm also looking at how to forbid PRs directly to the master branch to avoid this in the future.

**Comment by ypicchi-arm** [comment]
This last update should only fix the conflict in the copyright section. The other apparent changes should only be because of the rebase to the develop branch. Sorry for missing that earlier.

**Comment by paul-szczepanek-arm** [comment]
you can set the default branch to develop - that will make the PRs default to it

**Comment by markos** [comment]
good tip, thanks!

**Review by markos — APPROVED** [review]
LGTM, approved, there are a couple build failures with SIMDe, but they are known and unrelated ones.

**Pull request merged**